src/Akka.Streams.Kafka/Messages/CommittableOffsetBatch.cs

```diff
@@ -31,17 +32,18 @@ public CommittableOffsetBatch(IImmutableDictionary<GroupTopicPartition, OffsetAn
         Committers = committers;
         BatchSize = batchSize;
     }
-
-    /// <inheritdoc />
+    
     public long BatchSize { get; }
 
-    /// <inheritdoc />
-    public IImmutableSet<GroupTopicPartitionOffset> Offsets
+    /// INTERNAL COMMENT
+    /// Represents the offsets as they are, rather than how they're going to be committed.
+    ///
+    /// We have to +1 all offsets upon commit - which is what you get inside OffsetsAndMetadata.
+    public IImmutableDictionary<GroupTopicPartition, Offset> Offsets
     {
         get
         {
-            return OffsetsAndMetadata.Select(o => new GroupTopicPartitionOffset(o.Key, o.Value.Offset))
-                .ToImmutableHashSet();
+            return OffsetsAndMetadata.ToImmutableDictionary(c => c.Key, c => new Offset(c.Value.Offset.Value - 1L));
```

Comment:
Fixed most of our test assertions and this code also exists in the JVM - this is the "true" set of offsets that will be committed representing their current values. `OffsetsAndMetadata` represents the N+1 of each of these - and those are the values that need to be sent to the broker.